File: dsa-java/src/main/java/com/dsa/practice/algo/backtracking/PalindromePartitioning.java
package com.dsa.practice.algo.backtracking;

import java.util.ArrayList;
import java.util.List;

/**
 * Given a string s, partition s such that every
 * substring
 *  of the partition is a
 * palindrome
 * . Return all possible palindrome partitioning of s.
 *
 *
 *
 * Example 1:
 *
 * Input: s = "aab"
 * Output: [["a","a","b"],["aa","b"]]
 * Example 2:
 *
 * Input: s = "a"
 * Output: [["a"]]
 *
 *
 * Constraints:
 *
 * 1 <= s.length <= 16
 * s contains only lowercase English letters.
 */
public class PalindromePartitioning {

    /**
     * returns all the partitions that are palindrome
     * @param word whose palindrome partition we have to find
     * @return @{@link List} all the partitions that are palindrome
     */
    public List<List<String>> partition(String word) {
        List<List<String>> partitions = new ArrayList<>();
        List<String> currentPartition = new ArrayList<>();

        partitionHelper(word, 0, currentPartition, partitions);

        return partitions;
    }

    /**
     * recursive function to find all the partition that are palindrome
     * @param word whose palindrome partition we have to find
     * @param index from where partition needs to start
     * @param currentPartition partition which we are testing
     * @param partitions contains all the partition that are palindrome
     */
    private void partitionHelper(String word, int index, List<String> currentPartition, List<List<String>> partitions) {
        if (index == word.length()) {
            partitions.add(new ArrayList<>(currentPartition));
            return;
        }

        for (int endIndex = index; endIndex < word.length(); endIndex++) {
            if (isPalindrome(word, index, endIndex)) {
              currentPartition.add(word.substring(index, endIndex+1));
              partitionHelper(word, endIndex+1, currentPartition, partitions);
              currentPartition.removeLast();
            }
        }
    }

    /**
     * Checks whether the substring of the word is palindrome or not
     * @param word whose substring we have to check whether it is palindrome or not
     * @param startIndex start index of the substring
     * @param endIndex end index of the substring
     * @return
     */
    private boolean isPalindrome(String word, int startIndex, int endIndex) {
        while (startIndex <= endIndex) {

            if (word.charAt(startIndex) != word.charAt(endIndex)) {
                return false;
            }

            startIndex++;
            endIndex--;
        }
        return true;
    }

}

File: dsa-python/backtacking/palindrome_partitioning.py
from typing import List
import logging

# Configure logging
logging.basicConfig(level=logging.INFO, format='%(asctime)s - %(levelname)s - %(message)s')
logger = logging.getLogger(__name__)


class PalindromePartitioning:


    def partition(self, word: str) -> List[List[str]]:
        """
             Will found out all the partition of the word that are palindrome.

            word  String whose partitioning that are palindrome we have to see
        """
        resultant: List[List[str]] = []
        current_partition: List[str] = []

        logger.info(f"starting partitioning to find palindrome for word {word}")
        self._partitionHelper(word, current_partition, resultant, 0)

        return resultant

    def _partitionHelper(self, word, current_partition, resultant, start):
        """
        Using backtracking we are partitioning the word and finding whether the partition is palindrome or not.
        If not palindrome we are skipping else moving to another partition

        :param word: word whose palindrome partitioning we have to find
        :param current_partition: contains list of all the partition
        :param resultant: this is the final list which will contains all the palindrome partition
        :param start: index of the word from which we have to start
        :return:
        """
        if start == len(word):
            resultant.append(list(current_partition))
            return

        length: int = len(word)
        for end in range(start, length):
            if self.is_palindrome(word, start, end):
                word_substring = word[start: end + 1]
                logger.debug(f"adding { word_substring } to current partition")
                current_partition.append(word_substring)
                self._partitionHelper(word, current_partition, resultant, end + 1)
                current_partition.pop()

    def is_palindrome(self, word:str, start: int, end: int) -> bool:
        """
        Checks whether the given word is palindrome or not

        :param word: given string whose palindrome we have to check
        :param start: index of the word from which we have to check
        :param end:  end index of the word
        :return: bool
        """
        logger.debug(f"checking palindrome for string {word}")
        while start < end:
            if word[start] != word[end]:
                logger.debug(f"{word} is not a palindrome")
                return False
            start+=1
            end-=1

        logger.debug(f"{word} is a palindrome")
        return True

# Example usage
if __name__ == "__main__":
    print(PalindromePartitioning().partition("aab"))
    print(PalindromePartitioning().partition("aab"))